### ISSUE TITLE
Optional delay between the availability of Generator's result and the update of Kubernetes secret

### ISSUE THREAD

**Issue by beyondbill** [external-secrets/external-secrets#3114]
**Is your feature request related to a problem? Please describe.**
Our team is trying to automate CSP (cloud service providers such as AWS, Azure and GCP) key rotation with external-secrets + Vault Secret Engines. The keys are saved as Kubernetes secrets and consumed by application pods. Previously, key rotation would involve manual steps to create a new key from a CSP and update the Kubernetes secret with it. While external-secrets + Vault Secret Engines have demonstrated huge success in automating the manual key rotation in AWS and GCP, however, we're facing an issue with Azure. The Azure keys returned from Vault are instantly active only in the region where Vault is located in and could take up to 10 mins to be propagated globally. The Kubernetes secrets are immediately updated with the new Azure keys once they are available from Vault and application pods immediately takes the new keys from the secrets next time they are making requests to Azure. If a Azure API call is made to a region different from where the Vault that generated the Azure key is located in, it may keep failing in the next 10 min or so.

**Describe the solution you'd like**
I would like to add an optional delay between the availability of Generator's result and the update of Kubernetes secret, something like `ExternalSecret.spec.dataFrom.delayAfterDataAvailable`. It defaults to 0 and keeps backward compatibility if it's not used.

I think it's quite a common use case of external-secrets to automatically & periodically rotate keys in Kubernetes secrets. As an end user in this group, I would anticipate the still-valid old key in a Kubernetes secret not to be replaced with a new key that has not yet been globally valid, despite the key issuer has responded that back to external secrets. Sure the downstream services likely have re-trying and error handling to protect against a short period of invalid key. But that's not a guarantee and should never be taken for granted. The bottom line of key rotation is to always use valid keys. I'm not saying it's the duty of external secrets to validate the global consistency of the data returned from a Generator. But currently there's simply no way for a user to use external secrets automate the rotation of eventual-consistent keys except updating retrying and error handling in all the downstream services that use the key. Instead of affecting a potentially large amount of services, the proposed optional delay should be an easy win to boost user adoption

**Describe alternatives you've considered**
I have considered two alternatives, neither is feasible for our team though
- Increasing the CSP authentication retrying counts and period. But 10 min is just too long a time to be acceptable for our applications
- Having dedicated Azure key per region. It would require an enormous overhaul of downstream application code and over-complicate the authentication flow on Azure. It would also require Vault to be deployed per region, which is prohibitively costly.  

**Additional context**
I'm happy to dedicate some time this month to implement the proposed feature if the community feel it worths pursuing. Feel free to suggest where to start in the code or better alternative solutions!

**Comment by shuheiktgw** [comment]
Hi, @beyondbill 👋 Thanks for putting together the issue! From the description above, it was a bit unclear to me how each component interacts with the other, so could you describe the relationship between Azure, Vault, and ExternalSecret in more detail? 

A few random thoughts from me at this point:
- The root cause of this problem seems that ESO pulls changes from cloud providers instead of pulling them. Instead of adding this type of ad-hock field, pursuing https://github.com/external-secrets/external-secrets/issues/675 may be better.
- This looks like a problem specific to the Vault provider. Can we update the provider instead of adding a global field?
- `spec.dataFrom` does not seem the right place to add the field.

**Comment by beyondbill** [comment]
Thanks for the response @shuheiktgw ! In our use case, an `ExternalSecret` resource manages a Kubernetes secret that is used by applications to authenticate with Azure.
```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
...
spec:
  dataFrom:
    - sourceRef:
        generatorRef:
          apiVersion: generators.external-secrets.io/v1alpha1
          kind: VaultDynamicSecret
          name: azure-key-generator
  refreshInterval: 720h
  target:
    name: k8s-secret-with-azure-key-in-it
    ...
```
So every `720h` the external secret relies on VaultDynamicSecret `azure-key-generator` to call Vault's Azure Secret Engine, which in turn calls Azure API to request a new key. Azure API returns the new key to Vault, which in turn returns the key to external secret, which immediately update Kubernetes secret `k8s-secret-with-azure-key-in-it` with the new key.

The root of the problem we are facing is that this new key isn't globally ready yet in Azure. It takes time to propagate across all Azure regions, which could be up to 10 min. I don't think Azure will issue an event for global convergence reached. So a notification-based pulling mechanism as you suggested won't help this case.

The problem I described seems specific to Azure. But it could be generalized to any provider which may respond secret data back to external secret when the secret has not yet reached global convergence. That's why I believe a new field for optional delay between the secret data response and the update of Kubernetes secret should be placed in a generic location such as `spec.dataFrom`

**Comment by beyondbill** [comment]
@shuheiktgw Did you get a chance to review my comment above?

**Comment by shuheiktgw** [comment]
Sorry that I cannot get back to this issue 🙇 
After reviewing your comment, I remain unconvinced about adding the global field to the ExternalSecret spec. I believe the problem is too specific to the provider, so our initial step should be to investigate whether we can resolve it using VaultDynamicSecret. For instance, are there any methods available to programmatically determine if the provided key is globally propagated?

**Comment by beyondbill** [comment]
@shuheiktgw 
>  I remain unconvinced about adding the global field to the ExternalSecret spec. I believe the problem is too specific to the provider, so our initial step should be to investigate whether we can resolve it using VaultDynamicSecret.

Fair point. I'm fine with adding the proposed new option `delayAfterDataAvailable` to the specs of `VaultDynamicSecret`.

> are there any methods available to programmatically determine if the provided key is globally propagated?

Azure does not provide any API to show secret propagation status. All it tells users is to wait up to xxx minutes before using a newly generated secret. If we really want to check propagation status globally, we'll have to make a call to Azure API in every region and all have to succeed in order to prove the global propagation of the new secret is complete. I think a delay would be much easier to implement. And the delay works for any Vault secret engine (not just Azure) if we decide to add the new option of delay into `VaultDynamicSecret`.

**Comment by evrardj-roche** [comment]
Isn't the initial problem the propagation time of the provider?  Why should we care about this?
Alternatively, we can probably make it a flag of the controller, instead of adding it as a field of the ES spec... If you want to wait, I don't see a reason why some ES would be different than some others... Keep in mind you can also have multiple provider classes...

**Comment by alekc** [comment]
If you generate the secret locally with local generator and push it to the AzureVault via PushSecret, would not that solve the issue in this case?

**Comment by evrardj-roche** [comment]
@alekc if I got that right, the problem comes from the replication which is eventually consistent. If they remove replication they have to handle it themselves and it becomes _their_ workflow problem. I can see why users would not want that, as they are paying cloud providers for such service.

If I understand correctly, the problem comes whenever the cloud provider's replication is too slow for their business needs. For me, they have multiple choices (they are not exclusive to each other): Work with the cloud provider to improve performance within SLA (which is not a technical guarantee), handle the workflow themselves (as you suggested), be resilient in the face of eventual consistency. 
For me, the kubernetes is offering the latter (or the application on top of k8s should). We have tools for pushing the secrets if they want it. Users can contact Azure if performance is not matching their expectations/SLA.

Therefore, I don't think it's worth adding as a feature, personally, with the current use case.

Just highlighting something: "currently there's simply no way for a user to use external secrets automate the rotation of eventual-consistent keys except updating retrying and error handling in all the downstream services that use the key" -> Isn't that the cloud native way ? Trying to fix something that's not native by working on the infrastructure is quite an anti-pattern for long term resilience, no?

**Comment by alekc** [comment]
Yes, thats what I am saying. If I am reading this issue right, it is caused by the fact that the rotation happens on Azure side via the generator. and its replicated with delay across different regions. If the rotation happened on ESO side, it would push the rotated secret to the Azure, and synced up. 

Granted, I do not have any insight into what kind of secret it is, so I just left the comment above as an additional food for thought for anyone encountering the same issue (and yes, I do agree that its not really our place to wire this to accomodate 1 specific provider).

**Comment by evrardj-roche** [comment]
Hehe, we definitely agree :) Thanks for the comment!

My comment was also as "food for thought", just focusing on different difficulties: Organizational or on the principles.

I think with all of this, people will have enough tools to find the best solutions for them.

Let's see what people think.
